pkg/controller/atlasdeployment/serverless_private_endpoint.go

```diff
@@ -1,459 +1,271 @@
+// reconciliation. See validate.serverlessPrivateEndpoints
+func sortTasks(
+	privateEndpoints []akov2.ServerlessPrivateEndpoint,
+	atlasPrivateEndpoints []admin.ServerlessTenantEndpoint,
+) ([]akov2.ServerlessPrivateEndpoint, []akov2.ServerlessPrivateEndpoint, []admin.ServerlessTenantEndpoint) {
+	toCreate := make([]akov2.ServerlessPrivateEndpoint, 0, len(privateEndpoints))
+	toUpdate := make([]akov2.ServerlessPrivateEndpoint, 0, len(privateEndpoints))
+	toDelete := make([]admin.ServerlessTenantEndpoint, 0, len(atlasPrivateEndpoints))
+
+	privateEndpointsByName := map[string]*akov2.ServerlessPrivateEndpoint{}
+	for i := range privateEndpoints {
+		privateEndpoint := privateEndpoints[i]
+		privateEndpointsByName[privateEndpoint.Name] = &privateEndpoint
+	}
+
+	atlasPrivateEndpointsByName := map[string]*admin.ServerlessTenantEndpoint{}
+	for i := range atlasPrivateEndpoints {
+		atlasPrivateEndpoint := atlasPrivateEndpoints[i]
+		atlasPrivateEndpointsByName[atlasPrivateEndpoint.GetComment()] = &atlasPrivateEndpoint
+	}
+
+	// Collect all endpoints to create and update (connect)
+	for i := range privateEndpoints {
+		privateEndpoint := privateEndpoints[i]
+		atlasPrivateEndpoint, ok := atlasPrivateEndpointsByName[privateEndpoint.Name]
+
+		// If a private endpoint with a given name doesn't exist on Atlas, add to creation list
+		if !ok {
+			toCreate = append(toCreate, privateEndpoint)
 		}
-		if !found {
-			result.appendToCreate(desiredPE)
+
+		// If a private endpoint with a given name exists on Atlas with status RESERVED, add to update list (need to be connected)
+		if isReadyToConnect(&privateEndpoint, atlasPrivateEndpoint) {
+			toUpdate = append(toUpdate, privateEndpoint)
 		}
 	}
 
-	for _, existingPE := range existingPEs {
-		toDelete := true
-		for _, desiredPE := range result.PEToConnect {
-			if existingPE.ID == desiredPE.ID {
-				toDelete = false
-				break
-			}
-		}
-		for _, desiredPE := range result.PEToUpdateStatus {
-			if existingPE.ID == desiredPE.ID {
-				toDelete = false
-				break
-			}
-		}
-		if toDelete {
-			result.PEToDelete = append(result.PEToDelete, existingPE.ID)
+	for _, atlasPrivateEndpoint := range atlasPrivateEndpoints {
+		// If an existing Atlas private endpoint is not present in kubernetes resource, add to deletion list
+		if _, ok := privateEndpointsByName[atlasPrivateEndpoint.GetComment()]; !ok {
+			toDelete = append(toDelete, atlasPrivateEndpoint)
 		}
 	}
 
-	return &result
+	return toCreate, toUpdate, toDelete
 }
 
-func isReadySPEEqual(existingPE mongodbatlas.ServerlessPrivateEndpointConnection, desiredPE akov2.ServerlessPrivateEndpoint) bool {
-	return existingPE.Comment == desiredPE.Name && desiredPE.CloudProviderEndpointID == existingPE.CloudProviderEndpointID && desiredPE.PrivateEndpointIPAddress == existingPE.PrivateEndpointIPAddress
-}
+func mapStatuses(atlasPrivateEndpoints []admin.ServerlessTenantEndpoint) map[string]status.ServerlessPrivateEndpoint {
```

Comment:
rename this to `statusMap`, also note that plural of "status" is "status", see https://www.grammarbook.com/blog/singular-vs-plural/plural-of-status